discovery_engine_core/web-api2/src/db.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use secrecy::{ExposeSecret, Secret};
+use serde::{Deserialize, Serialize};
+use sqlx::{pool::PoolOptions, postgres::PgConnectOptions, Pool, Postgres};
+
+use crate::utils::serialize_redacted_opt;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Config {
+    /// The default base url.
+    #[serde(default = "default_base_url")]
```

Comment:
:question: What is the point in having all this default here and in the es part? We can easily have a configuration file that we use for testing.

Reply:
The cases are massively different. The engine is a library where the configuration is composed of parameters for the AI that we want to change mainly for A/B testing. Having a default configuration file would have meant deploying it alongside the library and having the client giving as the path to it making it more difficult for little gain.

Here having a file alongside the binary is pretty easy and there are no sane default values that can be used in production to connect to the database and elasticsearch.

If the automatic system for env vars is not good enough we can set some manually (PREFIX_DB_URL, PREFIX_DB_PASSWORD) if needed for the infrastructure otherwise we can drop that part.
For the priority of the options seems that cli > env > configuration has the most consensus on the internet but we can invert configuration and env. The important part is that we document it.

Reply:
> no sane default values that can be used in production

We have sane default values we do not change in production for **most** of the config fields, the fields we change to different values depending on production, testing etc. are the minority, by far.

> automatic system for env vars is not good

It is good enough, but if you want to have default values for each config field in the binary then you have to put default values on each config field which is what that code is doing. I don't understand why you have a problem with that.

>  the options seems that cli > env > configuration has the most consensus o

that is not what I mean, what I mean is with local override fiels like `config.test.toml` or similar which should be treated with more priority then e.g. some default `.env` but with less priority then e.g. `.env.local` or manually set env variables. 

The reason why you might want to have a override file is that having a config in tree and changing it for local testing is very error pone. So you want a solution without override files. But juggling multiple config files e.g. one not commited for local testing, one committed for testing without your changes, one with production defaults (or similar) can also get nasty especially the more configs there are in the files. So I would want to have as little configs in eachf file as possible especially given that most rarely change and are most times independent of the context. Which means having defaults in the program, which is seems to also be industry consensus given that this is how most programs behave.

Having default in the program also allows you to e.g. use a different default logging level for debug builds.

>  file alongside the binary is pretty easy 

it's not much easier then having it in a app, in both cases it's pretty trivial

Reply:
Two maybe relevant things:

- we should be able to derive `Default` based on `serde`
- there is an open issue/in-progress PR for `default_value` which would make it _way_ more readable/concise

Reply:
:pushpin: I noted this discussion down in a followup ticket.